auth/impersonate/impersonate.go

```diff
@@ -0,0 +1,211 @@
+				Audience: defaultAud,
+				Scopes:   []string{defaultScope},
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+	} else if opts.TokenProvider != nil {
+		client = internal.CloneDefaultClient()
+		if err := httptransport.AddAuthorizationMiddleware(client, opts.TokenProvider); err != nil {
+			return nil, err
+		}
+	} else {
+		client = opts.Client
+	}
+
+	// If a subject is specified a different auth-flow is initiated to
+	// impersonate as the provided subject (user).
+	if opts.Subject != "" {
+		return user(opts, client, lifetime, isStaticToken)
+	}
+
+	its := impersonatedTokenProvider{
+		client:          client,
+		targetPrincipal: opts.TargetPrincipal,
+		lifetime:        fmt.Sprintf("%.fs", lifetime.Seconds()),
+	}
+	for _, v := range opts.Delegates {
+		its.delegates = append(its.delegates, formatIAMServiceAccountName(v))
+	}
+	its.scopes = make([]string, len(opts.Scopes))
+	copy(its.scopes, opts.Scopes)
+
+	var tpo *auth.CachedTokenProviderOptions
+	if isStaticToken {
+		tpo = &auth.CachedTokenProviderOptions{
+			DisableAutoRefresh: true,
+		}
+	}
+	return auth.NewCachedTokenProvider(its, tpo), nil
+}
+
+// CredentialOptions for generating an impersonated credential token.
+type CredentialOptions struct {
+	// TargetPrincipal is the email address of the service account to
+	// impersonate. Required.
+	TargetPrincipal string
+	// Scopes that the impersonated credential should have. Required.
+	Scopes []string
+	// Delegates are the service account email addresses in a delegation chain.
+	// Each service account must be granted roles/iam.serviceAccountTokenCreator
+	// on the next service account in the chain. Optional.
+	Delegates []string
+	// Lifetime is the amount of time until the impersonated token expires. If
+	// unset the token's lifetime will be one hour and be automatically
+	// refreshed. If set the token may have a max lifetime of one hour and will
+	// not be refreshed. Service accounts that have been added to an org policy
+	// with constraints/iam.allowServiceAccountCredentialLifetimeExtension may
+	// request a token lifetime of up to 12 hours. Optional.
+	Lifetime time.Duration
+	// Subject is the sub field of a JWT. This field should only be set if you
+	// wish to impersonate as a user. This feature is useful when using domain
+	// wide delegation. Optional.
+	Subject string
+
+	// TokenProvider is the provider of the credentials used to fetch the ID
+	// token. If not provided, and a Client is also not provided, credentials
+	// will try to be detected from the environment. Optional.
+	TokenProvider auth.TokenProvider
+	// Client configures the underlying client used to make network requests
+	// when fetching tokens. If provided the client should provide it's own
+	// credentials at call time. Optional.
+	Client *http.Client
+}
+
+func formatIAMServiceAccountName(name string) string {
+	return fmt.Sprintf("projects/-/serviceAccounts/%s", name)
+}
+
+type generateAccessTokenReq struct {
```

Comment:
Nit: Other structs use "Request" in their names.

Reply:
done.